doc/Message-Protocol.md

```diff
@@ -1,111 +1,129 @@
+
+### Bit vector control flags
+Bits vectors are usually run-length encoded unless the raw unencoded
+bit vector is more compact than the rle-encoded representation. The
+rle-encoded flag indicates if the bit vector is encoded or raw.
+
+The rle-complement flags indicates if the complement of the bit vector
+was was encoded. The bit vector complement is encoded if this results
+in a more compact encoding. This flag is only useful with run-length
+encoding.
+
+~~~
+bit-vector-flags = &(
+  rle-encoded: 1,
+  rle-complement: 2
+)
+~~~
+
 ### Subscription flags
-~~~~py
+
+~~~
 sub-flags = &(
   delta: 1,        ; # indicate interests is a delta
   mute: 2          ; # mute has been indicated
 )
-~~~~
-### Header field member keys.
-For compactness the member keys are encoded as integers as listed below.
-~~~~py
-header-field = (
-  ? 1 => uint,        ; # port number sender is listening on
-  ? 2 => int,         ; # ttl - time to live in seconds
-  ? 3 => uuid,        ; # pub-id - unique indentifier for a publication
-  ? 4 => uint,        ; # seq-num - sequence number for a publication
-  ? 5 => bool,        ; # ack-req - indicates if an publisher is requesting an acknowledgement
-  ? 6 => bit-vector,  ; # bloom-filter -the bloom filter for a publication
-  ? 7 => sub-flags,   ; # sub-flags - indicates delta or mute
-  ? 8 => uuid,        ; # mesh-id - the mesh ID
-  ? 9 => bit-vector,  ; # needs - the needs bit vector
-  ? 10 => bit-vector  ; # interests - the interests bit vector
-)
-~~~~
-## DPS message types
+~~~
 
-### Publication message encoding
-The encoding of a publication.
-~~~~py
-pub = [
-  version: 1,
-  type: 1,
-  headers: { * header-field }, ; # port, ttl
-  body: { * header-field },    ; # ttl, pub-id, seq-num, ack-req, bloom-filter
-  payload: [
-    topics: [ + topic: tstr ]
-    payload: bstr
-  ]
-]
-~~~~
-### Subscription message encoding
-The encoding of a subscription.
-~~~~py
-sub = [
-  version: 1,
-  type: 2,
-  headers: { * header-field }, ; # port, seq-num
-  body: { * header-field }     ; # sub-flags, mesh-id, needs, interests or empty for unlink
-]
-~~~~
-### Acknowledgement message encoding
-The encoding of an acknowledgement message.
-~~~~py
-ack = [
-  version: 1,
-  type: 3,
-  body: { * header-field },    ; # pub-id, seq-num
-  payload: bstr
-]
-~~~~
-### Subscription acknowledgement message encoding
-The encoding of a subscription acknowledgement message.
-~~~~py
-sak = [
-  version: 1,
-  type: 4,
-  headers: { * header-field }  ; # port, seq-num
-]
-~~~~
+## Publication message
+
+~~~
+pub = 1
+~~~
+
+*port* and *ttl* are mandatory in the *unprotected* section.
+
+*ttl*, *pub-id*, *seq-num*, *ack-req* and *bloom-filter* are
+mandatory in the *protected* section.
+
+*topics* and *data* are mandatory in the *encrypted* section.
+
+## Subscription message
+
+~~~
+sub = 2
+~~~
+
+*port* and *seq-num* are mandatory in the *unprotected* section.
+
+In a regular subscription message, *sub-flags*, *mesh-id*, *needs*
```

Comment:
This doesn't look right to me. There is no keying material for sub/sub-acks so the protected section should be empty unless you are thinking of adding support for a mesh-global encryption key for authenticating sub/sub-ack messages. Previous thinking was that sub/sub-acks would be adequately protected by the node-to-node TLS links.

Reply:
Agreed, I moved them to the unprotected section in commit 35e1f780cad315ae43f7ad5b8254172e17a7adbb